src/kern/npf_alg_pptp.c

```diff
@@ -0,0 +1,788 @@
+		}
+
+		npfa_pptp_gre_slot_free(npc->npc_ctx, gre_slot, o_addr,
+				  npc->npc_ips[NPF_SRC]);
+		npfa_pptp_tcp_conn_unlock(gre_conns);
+		break;
+
+	case PPTP_WAN_ERROR_NOTIFY:
+		if (pptp->len < sizeof(pptp_msg_hdr_t))
+			return false;
+
+		npfa_pptp_tcp_conn_lock(gre_conns);
+		gre_slot = npfa_pptp_gre_slot_lookup_with_client_call_id(gre_conns,
+				  pptp->call_id);
+		if (gre_slot == NULL) {
+			npfa_pptp_tcp_conn_unlock(gre_conns);
+			return false;
+		}
+
+		orig_client_call_id = gre_slot->orig_client_call_id;
+		npfa_pptp_tcp_conn_unlock(gre_conns);
+
+		/* rewrite */
+		old_call_id = pptp->call_id;
+		pptp->call_id = orig_client_call_id;
+		tcp->check = npf_fixup16_cksum(tcp->check, old_call_id,
+				  orig_client_call_id);
+		break;
+
+	default:
+		return false;
+	}
+
+	return true;
+}
+
+/*
+ * Destroy PPTP TCP nat argument.
+ * It will expire all associated gre connections.
+ */
+static void
+npfa_pptp_tcp_destroy(npf_t *npf, npf_conn_t *con)
+{
+	pptp_gre_con_slot_t *gre_slot;
+	pptp_gre_conns_t *gre_conns;
+	npf_connkey_t *fw;
+	npf_addr_t ips[2];
+	uint16_t ids[2];
+	uint16_t alen;
+	uint16_t proto;
+
+	gre_conns = (pptp_gre_conns_t *)npf_nat_get_alg_arg(con->c_nat);
+	fw = npf_conn_getforwkey(con);
+
+	/* only ipv4 is supported */
+	if (NPF_CONNKEY_ALEN(fw) == sizeof(uint32_t) || gre_conns == NULL)
+		return;
+
+	npf_connkey_getkey(fw, &proto, ips, ids, &alen);
+
+	for (int i = 0; i < PPTP_MAX_GRE_PER_CLIENT; i++) {
+		gre_slot = &gre_conns->gre_con_slots[i];
+		if ((gre_slot->flags & PPTP_ALG_FL_FREE_SLOT) == 0)
+			npfa_pptp_gre_slot_free(npf, gre_slot, &ips[NPF_SRC], &ips[NPF_DST]);
+	}
+
+	npfa_pptp_gre_conns_fini(gre_conns);
+}
+
+/*
+ * PPTP GRE ALG translator.
+ */
+static bool
+npfa_pptp_gre_translate(npf_cache_t *npc, npf_nat_t *nt, bool forw)
+{
+	nbuf_t *nbuf = npc->npc_nbuf;
+	pptp_gre_hdr_t *gre;
+	pptp_gre_con_slot_t gre_slot;
+
+	if (forw || !npf_iscached(npc, NPC_IP4 | NPC_ALG_PPTP_GRE) ||
+			  npc->npc_proto != IPPROTO_GRE)
+		return false;
+
+	nbuf_reset(nbuf);
+	gre = nbuf_advance(nbuf, npc->npc_hlen, sizeof(pptp_gre_hdr_t));
+	if (gre == NULL)
+		return false;
+
+	gre_slot.u64 = (uint64_t)npf_nat_get_alg_arg(nt);
+	KASSERT(gre->call_id == gre_slot.ctx.client_call_id);
+	gre->call_id = gre_slot.orig_client_call_id;
+
+	return true;
+}
+
+/*
+ * Destroy gre connection.
+ * Put translated call-id back to the portmap.
+ */
+static void
+npfa_pptp_gre_destroy(npf_t *npf, npf_conn_t *con)
+{
+	pptp_gre_con_slot_t gre_slot;
+	npf_nat_t *nt;
+	npf_connkey_t *fw;
+	npf_addr_t ips[2];
+	uint16_t ids[2];
+	uint16_t alen;
+	uint16_t proto;
+
+	fw = npf_conn_getforwkey(con);
+	nt = con->c_nat;
+
+	/* only ipv4 is supported */
+	if (NPF_CONNKEY_ALEN(fw) != sizeof(uint32_t) || nt == NULL)
+		return;
+
+	npf_connkey_getkey(fw, &proto, ips, ids, &alen);
+
+	gre_slot.u64 = (uint64_t)npf_nat_get_alg_arg(nt);
+	if (gre_slot.ctx.client_call_id != 0)
+		npfa_translated_call_id_put(&ips[NPF_DST], gre_slot.ctx.client_call_id);
```

Comment:
Just thinking about possible races.. I suppose `npf_nat_set_alg_arg()` is not necessary here, right?

Reply:
Why would you need npf_nat_set_alg_arg() here? The code is executed when a connection is out of the scope of any worker threads and about to be destroyed, therefore there is no possible race.